### PR TITLE
Make immediately indexed array literals `@nogc`

### PR DESCRIPTION
**Pull request by dkorpel**
Start simple, doing this on multi-dimensional arrays is a lot more complex.

### CODE REVIEW

**Comment by rikkimax** [comment]
Does this work with intermediaries like enum declarations?

File: compiler/test/runnable/nogc.d [hunk 1]
```diff
@@ -68,12 +68,25 @@ void test12936() @nogc
 
 /***********************/
 
+void testIndexedArrayLiteral() @nogc
+{
+    int i = 2;
+    int x = [10, 20, 30, 40][i];
+    assert(x == 30);
+
+    enum arr = [100, 200, 300, 400][1 .. $];
```
Comment (thewilsonator) [inline compiler/test/runnable/nogc.d:77]: @rikkimax yes

**Review by thewilsonator — COMMENTED** [review]
this should check for ref
```d
ref int getFormat(int channels)
 {
     return [0, 1, 2, 3, 4][channels];
 }

...
writeln(getFormat(2));
```

which currently works and I presume would leak stack memory with this change.

**Comment by thewilsonator** [comment]
LDC does a GC to stack promotion optimisation. It doesn't set `@nogc` but it has otherwise the same effect. If you can do it easily in DMD, sure go for it, but if not, no great loss.

**Comment by dkorpel** [comment]
This probably requires a solution similar to the current `func.tookAddressOf` mechanism and https://github.com/dlang/dmd/pull/7981, which is for another day.

Interestingly though, no tests or buildkite projects broke with this PR, so I'll add a test for that.

**Pull request closed without merging**